# Import des bibliothèques

In [196]:
import re
from functools import reduce

# Lecture du fichier

In [197]:
with open ("real2.txt") as file:
    fichier = file.read()
fichier

"' mul(382,128)select(){*who(710,947)mul(117,325)?$#from()/select()mul(829,251)}@mul(17,183)(:*when()}?+,what()mul(911,142)[:>)who(824,820)/mul(199,484) when()mul(325,240))select()$~?'from(387,163)>what()@mul(920,723)where()}(#what()<who() !mul(274,899)what()mul(664,836)}~who();how()who(537,994)/(?]mul(257,635)^when()don't()!what()}where()#:when()>'where()mul(360,222),when()@&<^mul(268,245){{)%:from()<#mul(936,776)$select()!mul(474,825)how()}~mul(484,39)!?:@[*<mul(357,805)how()mul(261,810) {$>mul(306,422)$when()when()/@$$!mul(944,563)<!%!(from()mul(47,642)#^(why()(}:$mul(403,781)mul(382,778)$%$-)mul(48,400)@{?select()-/%when())}mul(114,537)^$&{&select()+why(){/mul(688,466) ):mul(950,333)when()}what()when(){>'!%+mul(974,802)what(291,78)*mul(394,250)why()<;mul(271,377)how()%**@who()*from()mul(569,753))who()*^mul(83,470)#-{from()when()-;mul(678,845)where()><;]$do()>what()(#)how() mul(958,48)select()from()<%mul(965,566)!};<where()mul(926,836)>*when()?)%do()}mul(465,920)>$%$when()mul(905,94

# On recupère les mul(...,...)

In [198]:
list_mult = [[int(nb1)*int(nb2),mul.start()]
             for mul in re.finditer(r'mul\((\d+),(\d+)\)', fichier) # récupère les mul(512,255)...
             for nb1,nb2 in [mul.groups()] # récupère les nombres contenus dans mul(622,262)
            ]
# nb1 * nb2 | start position
list_mult

[[48896, 2],
 [38025, 36],
 [208079, 66],
 [3111, 80],
 [129362, 110],
 [96316, 139],
 [78000, 158],
 [665160, 204],
 [246326, 240],
 [555104, 258],
 [163195, 299],
 [79920, 357],
 [65660, 380],
 [726336, 405],
 [391050, 427],
 [18876, 446],
 [287385, 464],
 [211410, 481],
 [129132, 497],
 [531472, 527],
 [30174, 550],
 [314743, 573],
 [297196, 585],
 [19200, 602],
 [61218, 635],
 [320608, 668],
 [316350, 683],
 [781148, 720],
 [98500, 745],
 [102167, 764],
 [428457, 797],
 [39010, 817],
 [572910, 845],
 [45984, 889],
 [546190, 916],
 [774136, 939],
 [427800, 967],
 [854320, 989],
 [577116, 1006],
 [48983, 1037],
 [299205, 1060],
 [380029, 1082],
 [461481, 1105],
 [107616, 1130],
 [42435, 1165],
 [89271, 1183],
 [19992, 1205],
 [191352, 1236],
 [33390, 1278],
 [635194, 1298],
 [490140, 1310],
 [618469, 1337],
 [49984, 1366],
 [80080, 1392],
 [25064, 1404],
 [285032, 1429],
 [453096, 1441],
 [695267, 1468],
 [68562, 1487],
 [76400, 1517],
 [2827, 1556],
 [699130, 1567],
 [273726, 1580],

# On récupère les do et don't

In [199]:
do_dont = ([["do",do.start()] for do in re.finditer(r'do\(\)',fichier)]+
          [["dont",dont.start()] for dont in re.finditer(r"don\'t\(\)",fichier)])

# On trie par ordre de position du fichier source
do_dont = sorted(do_dont,key=lambda x:x[1])

do_dont

[['dont', 318],
 ['do', 869],
 ['do', 962],
 ['dont', 1528],
 ['dont', 1619],
 ['do', 1751],
 ['do', 1889],
 ['dont', 3012],
 ['dont', 3077],
 ['dont', 3253],
 ['dont', 3594],
 ['dont', 4329],
 ['do', 4351],
 ['do', 4379],
 ['do', 4609],
 ['dont', 4660],
 ['dont', 4755],
 ['dont', 4960],
 ['do', 5332],
 ['do', 5488],
 ['do', 5535],
 ['dont', 5760],
 ['do', 5887],
 ['do', 5968],
 ['do', 6056],
 ['dont', 6259],
 ['do', 6632],
 ['do', 7266],
 ['dont', 7612],
 ['dont', 8087],
 ['do', 8180],
 ['dont', 8437],
 ['dont', 8480],
 ['do', 8582],
 ['dont', 8855],
 ['do', 9456],
 ['do', 9802],
 ['do', 9901],
 ['dont', 10384],
 ['do', 10539],
 ['dont', 10599],
 ['do', 10643],
 ['do', 11032],
 ['dont', 11122],
 ['dont', 11165],
 ['dont', 11203],
 ['do', 11281],
 ['dont', 11451],
 ['do', 11492],
 ['do', 12149],
 ['dont', 12168],
 ['do', 12336],
 ['do', 12864],
 ['do', 13206],
 ['do', 13263],
 ['do', 13553],
 ['do', 13586],
 ['dont', 13828],
 ['dont', 14327],
 ['do', 14603],
 ['dont', 15184],
 ['dont',

In [ ]:
# On optimise la liste ---> si on a un do()...do()...dont() on simplifie en do()...dont()
# On recupere les indices à supprimer
prev_statut = do_dont[0][0]
pos_supp=[]


for statut,pos in do_dont[1:]:
    if prev_statut == statut:
        pos_supp.append(pos)
    prev_statut = statut
pos_supp

# On supprimes les élements redondants
clean_list = [liste[1] for liste in do_dont if liste[1] not in pos_supp]


# Output : dont:318
#          do:869.... 
#          do...
#          dont....

[318,
 869,
 1528,
 1751,
 3012,
 4351,
 4660,
 5332,
 5760,
 5887,
 6259,
 6632,
 7612,
 8180,
 8437,
 8582,
 8855,
 9456,
 10384,
 10539,
 10599,
 10643,
 11122,
 11281,
 11451,
 11492,
 12168,
 12336,
 13828,
 14603,
 15184,
 15283,
 15494,
 16253,
 17129,
 18083,
 18323,
 18431]

On a maintenant la liste des do/dont sans doublons successifs

In [212]:
# On regroupe le premier do avec le premier don't ect...
group_clean_list = list(zip(clean_list[::2],clean_list[1::2]))
group_clean_list

[(318, 869),
 (1528, 1751),
 (3012, 4351),
 (4660, 5332),
 (5760, 5887),
 (6259, 6632),
 (7612, 8180),
 (8437, 8582),
 (8855, 9456),
 (10384, 10539),
 (10599, 10643),
 (11122, 11281),
 (11451, 11492),
 (12168, 12336),
 (13828, 14603),
 (15184, 15283),
 (15494, 16253),
 (17129, 18083),
 (18323, 18431)]

On doit doit reparcourir le fichier source en supprimmant les multiplication entre la position 318 et 869 puis entre 1528 et 1751...

# On reparcourt le fichier

In [218]:
# On parcours la liste de multiplications
list_supp_pos_mult = []
for multiplication,pos_mult in list_mult:
    for intervalle in group_clean_list:
        # Si la position est comprise dans l'intervalle
        if intervalle[0] < pos_mult < intervalle[1]:
            list_supp_pos_mult.append(pos_mult)

sum(mult for mult,pos_mult in list_mult if pos_mult not in list_supp_pos_mult)

97529391